File: scripts/jnpf_auth.py
#!/usr/bin/env python3
"""
JNPF 无浏览器登录 + API 调用（与前端 MD5+AES 一致）

依赖: pip install requests pycryptodome

用法:
  python scripts/jnpf_auth.py login
  python scripts/jnpf_auth.py login --json
  python scripts/jnpf_auth.py GET /api/oauth/CurrentUser
  python scripts/jnpf_auth.py POST /api/studio/pipeline/execute/create '{"name":"t","userRequirement":"测试"}'

环境变量: JNPF_API_URL, JNPF_ACCOUNT, JNPF_PASSWORD, JNPF_CIPHER_KEY
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import sys
from pathlib import Path

try:
    import requests
    from Crypto.Cipher import AES
    from Crypto.Util.Padding import pad
except ImportError:
    print("请先安装: pip install requests pycryptodome", file=sys.stderr)
    sys.exit(1)

REPO_ROOT = Path(__file__).resolve().parent.parent
SESSION_FILE = REPO_ROOT / "scripts" / ".jnpf-session.json"

API_URL = os.environ.get("JNPF_API_URL", "http://localhost:5000").rstrip("/")
ACCOUNT = os.environ.get("JNPF_ACCOUNT", "admin")
PASSWORD = os.environ.get("JNPF_PASSWORD", "123456")
CIPHER_KEY = os.environ.get("JNPF_CIPHER_KEY", "EY8WePvjM5GGwQzn")
ORIGIN = os.environ.get("JNPF_ORIGIN", "pc")


def encrypt_password(plain: str, key: str = CIPHER_KEY) -> str:
    md5hex = hashlib.md5(plain.encode("utf-8")).hexdigest()
    cipher = AES.new(key.encode("utf-8"), AES.MODE_ECB)
    encrypted = cipher.encrypt(pad(md5hex.encode("utf-8"), AES.block_size))
    return encrypted.hex()


def jwt_exp_ms(token: str) -> int | None:
    try:
        payload = token.split(".")[1]
        padding = "=" * (-len(payload) % 4)
        data = json.loads(base64.urlsafe_b64decode(payload + padding))
        exp = data.get("exp")
        return int(exp) * 1000 if exp else None
    except Exception:
        return None


def load_session() -> dict | None:
    if not SESSION_FILE.exists():
        return None
    try:
        data = json.loads(SESSION_FILE.read_text(encoding="utf-8"))
        exp = data.get("expiresAt") or jwt_exp_ms(data.get("token", ""))
        if exp and __import__("time").time() * 1000 > exp - 60_000:
            return None
        return data
    except Exception:
        return None


def save_session(session: dict) -> None:
    SESSION_FILE.parent.mkdir(parents=True, exist_ok=True)
    SESSION_FILE.write_text(json.dumps(session, indent=2), encoding="utf-8")


def login(force: bool = False) -> dict:
    if not force:
        cached = load_session()
        if cached and cached.get("token") and cached.get("apiUrl") == API_URL:
            return cached

    password = encrypt_password(PASSWORD)
    resp = requests.post(
        f"{API_URL}/api/oauth/Login",
        data={
            "account": ACCOUNT,
            "password": password,
            "code": "",
            "timestamp": "",
            "origin": "password",
            "grant_type": "password",
        },
        headers={"Content-Type": "application/x-www-form-urlencoded", "jnpf-origin": ORIGIN},
        timeout=30,
    )
    body = resp.json()
    if body.get("code") != 200:
        raise RuntimeError(f"Login failed: HTTP {resp.status_code} {body}")

    token = body["data"]["token"]
    if token.startswith("Bearer "):
        token = token[7:].strip()
    session = {
        "apiUrl": API_URL,
        "account": ACCOUNT,
        "token": token,
        "expiresAt": jwt_exp_ms(token),
        "loginAt": __import__("datetime").datetime.utcnow().isoformat() + "Z",
    }
    save_session(session)
    return session


def api_request(method: str, path: str, body: dict | str | None = None, retry: bool = True) -> dict:
    session = login()
    url = path if path.startswith("http") else f"{API_URL}{path if path.startswith('/') else '/' + path}"
    headers = {"Authorization": f"Bearer {session['token']}", "jnpf-origin": ORIGIN}
    kwargs: dict = {"headers": headers, "timeout": 120}
    if body is not None:
        headers["Content-Type"] = "application/json"
        kwargs["json"] = body if isinstance(body, dict) else json.loads(body)

    resp = requests.request(method.upper(), url, **kwargs)
    if resp.status_code == 401 and retry:
        session = login(force=True)
        headers["Authorization"] = f"Bearer {session['token']}"
        resp = requests.request(method.upper(), url, **kwargs)

    try:
        data = resp.json()
    except Exception:
        data = resp.text
    return {"status": resp.status_code, "ok": resp.ok, "data": data}


def main() -> None:
    args = sys.argv[1:]
    if not args or args[0] in ("-h", "--help"):
        print(__doc__)
        sys.exit(0)

    if args[0] == "login":
        force = "--force" in args
        sess = login(force=force)
        if "--json" in args:
            print(json.dumps(sess, indent=2))
        else:
            print(sess["token"])
        return

    method = args[0].upper()
    path = args[1] if len(args) > 1 else ""
    if not path:
        print("缺少 path", file=sys.stderr)
        sys.exit(1)
    body = json.loads(args[2]) if len(args) > 2 else None
    result = api_request(method, path, body)
    print(json.dumps(result, indent=2, ensure_ascii=False))
    if not result["ok"]:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/lib/jnpf-auth.mjs
/**
 * JNPF 无浏览器登录 — MD5 + AES-ECB（与 PC/App 前端一致）
 *
 * 用法（模块）：
 *   import { login, encryptPassword, apiRequest } from './lib/jnpf-auth.mjs';
 *
 * 用法（CLI 取 token）：
 *   node scripts/lib/jnpf-auth.mjs
 *   node scripts/lib/jnpf-auth.mjs --json
 */

import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, '../..');
const SESSION_FILE = path.join(REPO_ROOT, 'scripts', '.jnpf-session.json');

const DEFAULTS = {
  apiUrl: process.env.JNPF_API_URL || 'http://localhost:5000',
  account: process.env.JNPF_ACCOUNT || 'admin',
  password: process.env.JNPF_PASSWORD || '123456',
  cipherKey: process.env.JNPF_CIPHER_KEY || 'EY8WePvjM5GGwQzn',
  origin: process.env.JNPF_ORIGIN || 'pc', // axios 默认 jnpf-origin: pc
};

/** 与 LoginForm.vue：encryptByMd5 → encryptByAES(useHex) */
export function encryptPassword(plainPassword, cipherKey = DEFAULTS.cipherKey) {
  const md5hex = crypto.createHash('md5').update(plainPassword, 'utf8').digest('hex');
  const key = Buffer.from(cipherKey, 'utf8');
  if (key.length !== 16) throw new Error(`AES key must be 16 bytes, got ${key.length}`);
  const cipher = crypto.createCipheriv('aes-128-ecb', key, null);
  const encrypted = Buffer.concat([cipher.update(md5hex, 'utf8'), cipher.final()]);
  return encrypted.toString('hex');
}

/** JNPF 返回的 token 可能已含 "Bearer " 前缀 */
export function normalizeToken(token) {
  if (!token) return '';
  return token.startsWith('Bearer ') ? token.slice(7).trim() : token.trim();
}

export function authHeader(token) {
  const t = normalizeToken(token);
  return t ? `Bearer ${t}` : '';
}

function decodeJwtExp(token) {
  try {
    const payload = token.split('.')[1];
    const json = JSON.parse(Buffer.from(payload, 'base64url').toString('utf8'));
    return typeof json.exp === 'number' ? json.exp * 1000 : null;
  } catch {
    return null;
  }
}

export function loadCachedSession() {
  try {
    if (!fs.existsSync(SESSION_FILE)) return null;
    const data = JSON.parse(fs.readFileSync(SESSION_FILE, 'utf8'));
    if (!data.token) return null;
    const expMs = data.expiresAt ?? decodeJwtExp(data.token);
    if (expMs && Date.now() > expMs - 60_000) return null;
    return data;
  } catch {
    return null;
  }
}

export function saveSession(session) {
  fs.mkdirSync(path.dirname(SESSION_FILE), { recursive: true });
  fs.writeFileSync(SESSION_FILE, JSON.stringify(session, null, 2), 'utf8');
}

/**
 * POST /api/oauth/Login → { token, ... }
 */
export async function login(options = {}) {
  const cfg = { ...DEFAULTS, ...options };
  const cached = options.force ? null : loadCachedSession();
  if (cached?.token && cached.apiUrl === cfg.apiUrl && cached.account === cfg.account) {
    return cached;
  }

  const password = encryptPassword(cfg.password, cfg.cipherKey);
  const body = new URLSearchParams({
    account: cfg.account,
    password,
    code: '',
    timestamp: '',
    origin: 'password',
    grant_type: 'password',
  });

  const res = await fetch(`${cfg.apiUrl.replace(/\/$/, '')}/api/oauth/Login`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'jnpf-origin': cfg.origin,
    },
    body: body.toString(),
  });

  const json = await res.json().catch(async () => ({ raw: await res.text() }));
  if (!res.ok || json.code !== 200) {
    throw new Error(`Login failed HTTP ${res.status} code=${json.code} msg=${json.msg || JSON.stringify(json)}`);
  }

  const token = normalizeToken(json.data?.token);
  if (!token) throw new Error('Login OK but no token in response');

  const session = {
    apiUrl: cfg.apiUrl,
    account: cfg.account,
    token,
    expiresAt: decodeJwtExp(token),
    loginAt: new Date().toISOString(),
  };
  saveSession(session);
  return session;
}

/** 带 Bearer 的 fetch；401 时可选重登一次；timeoutMs 默认 60s */
export async function apiRequest(method, urlPath, { body, token, session, retry = true, timeoutMs = 60_000 } = {}) {
  const cfg = { ...DEFAULTS };
  let sess = session || loadCachedSession();
  if (!sess?.token) sess = await login(cfg);
  if (!token) token = sess.token;

  const base = (sess.apiUrl || cfg.apiUrl).replace(/\/$/, '');
  const url = urlPath.startsWith('http') ? urlPath : `${base}${urlPath.startsWith('/') ? '' : '/'}${urlPath}`;

  const headers = {
    Authorization: authHeader(token),
    'jnpf-origin': cfg.origin,
  };
  let payload;
  if (body !== undefined) {
    headers['Content-Type'] = 'application/json';
    payload = typeof body === 'string' ? body : JSON.stringify(body);
  }

  const res = await fetch(url, {
    method,
    headers,
    body: payload,
    signal: timeoutMs ? AbortSignal.timeout(timeoutMs) : undefined,
  });
  if ((res.status === 401 || res.status === 600) && retry) {
    const fresh = await login({ ...cfg, force: true });
    return apiRequest(method, urlPath, {
      body,
      token: fresh.token,
      session: fresh,
      retry: false,
      timeoutMs,
    });
  }

  const text = await res.text();
  let json;
  try {
    json = JSON.parse(text);
  } catch {
    json = text;
  }
  return { ok: res.ok, status: res.status, json, text, headers: Object.fromEntries(res.headers.entries()) };
}

/** JNPF RESTfulResult：HTTP 200 但 code 可能非 200 */
export function isJnpfOk(result) {
  if (!result?.ok) return false;
  const j = result.json;
  if (j && typeof j === 'object' && 'code' in j) return j.code === 200;
  return true;
}

export function jnpfData(result) {
  const j = result?.json;
  if (j && typeof j === 'object' && 'data' in j) return j.data;
  return j;
}

/** 兼容 PascalCase / camelCase 字段 */
export function pick(obj, ...keys) {
  if (!obj || typeof obj !== 'object') return undefined;
  for (const k of keys) {
    if (obj[k] !== undefined) return obj[k];
  }
  return undefined;
}

// CLI: node jnpf-auth.mjs [--json] [--force]
const isMain = process.argv[1] && path.resolve(process.argv[1]) === path.resolve(fileURLToPath(import.meta.url));
if (isMain) {
  const force = process.argv.includes('--force');
  const asJson = process.argv.includes('--json');
  login({ force })
    .then(s => {
      if (asJson) console.log(JSON.stringify(s, null, 2));
      else console.log(s.token);
    })
    .catch(err => {
      console.error(err.message || err);
      process.exit(1);
    });
}
